libparsec/crates/platform_mountpoint/src/windows/mod.rs

```diff
@@ -0,0 +1,473 @@
+        .unwrap_or_default()
+}
+
+impl From<&EntryInfo> for FileInfo {
+    fn from(value: &EntryInfo) -> Self {
+        let created = filetime_from_utc(value.created);
+        let updated = filetime_from_utc(value.updated);
+
+        match value.ty {
+            EntryInfoType::File => *FileInfo::default()
+                .set_file_attributes(value.ty.into())
+                .set_creation_time(created)
+                .set_last_access_time(updated)
+                .set_last_write_time(updated)
+                .set_change_time(updated)
+                .set_index_number(value.id.as_u128() as u64)
+                .set_file_size(value.size)
+                .set_allocation_size(
+                    ((value.size + ALLOCATION_UNIT - 1) / ALLOCATION_UNIT) * ALLOCATION_UNIT,
+                ),
+            EntryInfoType::Dir => *FileInfo::default()
+                .set_file_attributes(value.ty.into())
+                .set_creation_time(created)
+                .set_last_access_time(updated)
+                .set_last_write_time(updated)
+                .set_change_time(updated)
+                .set_index_number(value.id.as_u128() as u64),
+            EntryInfoType::ParsecEntryInfo => *FileInfo::default()
+                .set_file_attributes(value.ty.into())
+                // Arbitrary non-zero size
+                .set_file_size(1024)
+                .set_time(filetime_now()),
+        }
+    }
+}
+
+impl From<EntryInfoType> for FileAttributes {
+    fn from(value: EntryInfoType) -> Self {
+        FileAttributes::not_content_indexed()
+            | match value {
+                EntryInfoType::File => FileAttributes::archive(),
+                EntryInfoType::Dir => FileAttributes::directory(),
+                EntryInfoType::ParsecEntryInfo => FileAttributes::archive(),
+            }
+    }
+}
+
+impl From<&EntryInfo> for FileAttributes {
+    fn from(value: &EntryInfo) -> Self {
+        value.ty.into()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Context {
+    File(File),
+    Dir(Dir),
+    Info(ParsecEntryInfo),
+}
+
+impl Context {
+    fn path(&self) -> &Path {
+        match self {
+            Self::File(file) => &file.path,
+            Self::Dir(dir) => &dir.path,
+            Self::Info(info) => &info.path,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct File {
+    path: PathBuf,
+    fd: FileDescriptor,
+}
+
+#[derive(Debug, Clone)]
+pub struct Dir {
+    path: PathBuf,
+}
+
+#[derive(Debug, Clone)]
+pub struct ParsecEntryInfo {
+    path: PathBuf,
+    encoded: String,
+}
+
+impl ParsecEntryInfo {
+    fn read(&self, buffer: &mut [u8], offset: usize) -> usize {
+        let end_offset = std::cmp::min(self.encoded.len(), offset + buffer.len());
+        let data = &self.encoded.as_bytes()[offset..end_offset];
+        buffer[..data.len()].copy_from_slice(data);
+        data.len()
+    }
+}
+
+impl FileSystemContext for MountpointManager {
+    type FileContext = Arc<Mutex<Context>>;
```

Comment:
It worked because behind the wrapper, I use Box before to pass to `winfsp`

Reply:
So after more reading on the WinFSP API, a little change on what winfsp-wrs should expose:

From [winfsp documentation](https://winfsp.dev/doc/WinFsp-API-winfsp.h/):
> Most of the operations accept a FileContext parameter. This parameter has different meanings depending on the value of the FSP_FSCTL_VOLUME_PARAMS flags UmFileContextIsUserContext2 and UmFileContextIsFullContext.
>
> There are three cases to consider:
>
>    When both of these flags are unset (default), the FileContext parameter represents the file node. The file node is a void pointer (or an integer that can fit in a pointer) that is used to uniquely identify an open file. Opening the same file name should always yield the same file node value for as long as the file with that name remains open anywhere in the system.
>
>    When the UmFileContextIsUserContext2 is set, the FileContext parameter represents the file descriptor. The file descriptor is a void pointer (or an integer that can fit in a pointer) that is used to identify an open instance of a file. Opening the same file name may yield a different file descriptor.
>
>    When the UmFileContextIsFullContext is set, the FileContext parameter is a pointer to a FSP_FSCTL_TRANSACT_FULL_CONTEXT. This allows a user mode file system to access the low-level UserContext and UserContext2 values. The UserContext is used to store the file node and the UserContext2 is used to store the file descriptor for an open file.

So long story short:
- winfsp-wrs has to deal with FileContext being either A) a pointer-sized scalar (i.e. a `u32` that correspond to a handle) or B) an actual pointer on data (in this case the data must not allow mutability given winfsp may call multiple operations at the same time with the same FileContext)
- winfsp-wrs may call close while another read/write operation is still ongoing. This means the data under the FileContext should not be freed in the close operation (typically this means if the FileContext contains a pointer on data, those data should be under a reference counter so that they get freed only when the read/write operation finishes)
- we need conversion operation to turn a FileContext pointer into the concrete type used in the implementation

In practice what we should have is:
```rust
use std::sync::Arc;

trait FileContextKind {
    // Write the data into winfsp's `PVOID *PFileContext`
    fn write(self, out: *mut libc::c_void);

    // write_ptr|scalar/full_context should be used depending of
    // UmFileContextIsUserContext2 and UmFileContextIsFullContext
    fn write_ptr<T>(ptr: *const T, out: *mut libc::c_void) {
        todo!()
    }
    fn write_scalar(val: usize, out: *mut libc::c_void) {
        todo!()
    }
//    fn write_full_context(context: FSP_FSCTL_TRANSACT_FULL_CONTEXT, out: *mut libc::c_void) {
//        todo!()
//    }

    // Retreive the data from winfsp's `FileContext`
    // If the data are Arc, the refcount should be increased
    // This is called in winfsp-wrs before calling FileSystemContext::read/write etc.
    fn access(raw: usize) -> Self;
    // Retreive the data from winfsp's `FileContext`
    // If the data are Arc, the refcount shouldn't be increased so that
    // they are freed once all operations on it have finished.
    // This is called in winfsp-wrs before calling FileSystemContext::close
    fn access_for_close(raw: usize) -> Self;
}

trait FileSystemContext<T> {
    type FileContext: FileContextKind;
    fn open(&self) -> Result<Self::FileContext, usize>;
    // winfsp-wrs does a `unsafe { &*ptr }` to convert the `*mut T` into a `&mut T`
    // (this is safe given the pointer is only freed when passing it to close)
    fn read(&self, file_context: Self::FileContext) -> Result<usize, usize>;
    // close must give back ownership of the pointer given it must now be freed
    fn close(&self, file_context: Self::FileContext);
}

// Example of implementation using an Arc

enum FileOrDir {
 File,
 Dir,
}

impl FileContextKind for Arc<FileOrDir> {
    fn write(self, out: *mut libc::c_void) {
      Self::write_ptr(Arc::into_raw(self), out);
    }
    fn access(raw: usize) -> Self {
        Self::access_for_close(raw).clone()
    }
    fn access_for_close(raw: usize) -> Self {
        unsafe{ Arc::<FileOrDir>::from_raw(raw as *const FileOrDir) }
    }
}


// Example using a scalar

impl FileContextKind for usize {
    fn write(self, out: *mut libc::c_void) {
      Self::write_scalar(self, out);
    }
    fn access(raw: usize) -> Self {
        raw
    }
    fn access_for_close(raw: usize) -> Self {
        raw
    }
}

```